tool/tsh/db.go

```diff
@@ -432,6 +417,64 @@ func maybeStartLocalProxy(cf *CLIConf, tc *client.TeleportClient, profile *clien
 	}, nil
 }
 
+// localProxyConfig is an argument pack used in prepareLocalProxyOptions().
+type localProxyConfig struct {
+	cliConf         *CLIConf
+	teleportClient  *client.TeleportClient
+	profile         *client.ProfileStatus
+	routeToDatabase *tlsca.RouteToDatabase
+	database        types.Database
+	listener        net.Listener
+}
+
+// prepareLocalProxyOptions created localProxyOpts needed to create local proxy from localProxyConfig.
+func prepareLocalProxyOptions(arg *localProxyConfig) (localProxyOpts, error) {
+	// If user requested no client auth, open an authenticated tunnel using
+	// client cert/key of the database.
+	certFile := arg.cliConf.LocalProxyCertFile
+	if certFile == "" && arg.cliConf.LocalProxyTunnel {
+		certFile = arg.profile.DatabaseCertPathForCluster(arg.cliConf.SiteName, arg.routeToDatabase.ServiceName)
+	}
+	keyFile := arg.cliConf.LocalProxyKeyFile
+	if keyFile == "" && arg.cliConf.LocalProxyTunnel {
+		keyFile = arg.profile.KeyPath()
+	}
+
+	opts := localProxyOpts{
+		proxyAddr: arg.teleportClient.WebProxyAddr,
+		listener:  arg.listener,
+		protocols: []common.Protocol{common.Protocol(arg.routeToDatabase.Protocol)},
+		insecure:  arg.cliConf.InsecureSkipVerify,
+		certFile:  certFile,
+		keyFile:   keyFile,
+	}
+
+	// For SQL Server connections, local proxy must be configured with the
+	// client certificate that will be used to route connections.
+	if arg.routeToDatabase.Protocol == defaults.ProtocolSQLServer {
+		opts.certFile = arg.profile.DatabaseCertPathForCluster("", arg.routeToDatabase.ServiceName)
+		opts.keyFile = arg.profile.KeyPath()
+	}
+
+	// To set correct MySQL server version DB proxy needs additional protocol.
+	if !arg.cliConf.LocalProxyTunnel && arg.routeToDatabase.Protocol == defaults.ProtocolMySQL {
+		if arg.database == nil {
+			var err error
+			arg.database, err = getDatabase(arg.cliConf, arg.teleportClient, arg.routeToDatabase.ServiceName)
+			if err != nil {
+				return localProxyOpts{}, trace.Wrap(err)
+			}
+		}
+
+		mysqlServerVersionProto := mySQLVersionToProto(arg.database)
+		if mysqlServerVersionProto != "" {
+			opts.protocols = append(opts.protocols, common.Protocol(mysqlServerVersionProto))
+		}
+	}
```

Comment:
I think we are getting to a point where we should extract those methods closer to the DB logic we have for each Protocol
Maybe not with this PR, but if we add another `if ... protocolX` I thnik we should re-write this flow